src/lib/bcmp/bm/bm_l2.cpp

```diff
@@ -354,8 +357,24 @@ err_t bm_l2_rx(void* device_handle, uint8_t* payload, uint16_t payload_len, uint
 err_t bm_l2_link_output(struct netif *netif, struct pbuf *pbuf) {
     (void) netif;
 
-    /* Send on all available ports (Multicast) */
-    return bm_l2_tx(pbuf, bm_l2_ctx.available_ports_mask);
+    // by default, send to all ports
+    uint8_t port_mask = bm_l2_ctx.available_ports_mask;
+
+    // if the application set an egress port, send only to that port
+    constexpr size_t bcmp_egress_port_offset_in_dest_addr = 13;
+    const size_t egress_port_idx = sizeof(struct eth_hdr) + offsetof(struct ip6_hdr, dest) +
+                                   bcmp_egress_port_offset_in_dest_addr;
+    uint8_t *eth_frame = static_cast<uint8_t *>(pbuf->payload);
+    uint8_t egress_port = eth_frame[egress_port_idx];
+    constexpr uint8_t num_adin_ports = 2; // TODO generalize
+    if (egress_port > 0 && egress_port <= num_adin_ports) {
+      port_mask = 1 << (egress_port - 1);
+    }
+
+    // clear the egress port set by the application
+    eth_frame[egress_port_idx] = 0;
+
+    return bm_l2_tx(pbuf, port_mask);
 }
```

Comment:
This is the callback at layer 2, called by LWIP after the ethernet frame header has been added. This:

1. inspects the egress port added to the destination IP address by `bcmp_ll_forward`
2. uses that to build a port mask
3. clears those bits in the destination IP address so that we send to the actual multicast group that other nodes are subscribed to
4. and sends the packet with the port mask to the ADIN2111 via `bm_l2_tx`.